### ISSUE TITLE
Race condition in writing config file when configure run immediately after new instance launch

### ISSUE THREAD

**Issue by matt-richardson** [OctopusDeploy/Issues#3456]
If you install and start a server using the exact steps that the setup wizard generates, and then immediately follow them by the exact steps that the ssl bindings wizard generates, there is a race condition between the initial launch (where it moves settings to the database) and the `configure` command writing to the config file.

```
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" create-instance --instance "Primary" --config "C:\OctopusTest\Octopus-Primary\OctopusServer-Primary.config"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" configure --instance "Primary" --home "C:\OctopusTest\Octopus-Primary" --storageConnectionString "Data Source=(local)\SQLEXPRESS;Initial Catalog=OctopusDeploy-TestPrimary;Integrated Security=True" --upgradeCheck "False" --upgradeCheckWithStatistics "False" --webAuthenticationMode "UsernamePassword" --webForceSSL "False" --webListenPrefixes "http://localhost:8065/" --commsListenPort "11000" --serverNodeName "BOB" --commsListenWebSocket "https://+:24500/WebSocketComms/Primary"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" database --instance "Primary" --delete --create --grant "NT AUTHORITY\SYSTEM"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" service --instance "Primary" --stop
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" admin --instance "Primary" --username "admin" --email "" --password "Supers3cretPassw0rd!"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" service --instance "Primary" --install --reconfigure --start --dependOn "MSSQL$SQLEXPRESS"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" service --stop --instance "Primary"
"netsh.exe" http delete sslcert ipport=0.0.0.0:18065
"netsh.exe" http add sslcert ipport=0.0.0.0:18065 appid={E2096A4C-2391-4BE1-1D3C-E353F930E7F1} certhash=3C1B3A254501947C9AB3D170E03FDCCD0803EF93 certstorename=My
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" configure --instance "Primary" --webForceSSL "False"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" configure --instance "Primary" --webListenPrefixes "http://localhost:8065/,https://localhost:18065/"
"D:\Data\Dev\Octopus\OctopusDeploy\source\Octopus.Server\bin\Octopus.Server.exe" service --instance "Primary" --stop --start
```

Usually, its okay, but occasionally the changes the `configure` command makes are overwritten by the `move to database` changes.

This was found in the API tests where it launched the commands using `OctopusSetupWizardModel` and then immediately ran the commands from `WebBindingsWizardModel`.

### main thread from server startup:

```
 at Octopus.Shared.Configuration.XmlFileKeyValueStore.OpenForWriting() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Configuration\XmlFileKeyValueStore.cs:line 41
   at Octopus.Shared.Configuration.XmlKeyValueStore.SaveSettings(IDictionary`2 settingsToSave) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Configuration\XmlKeyValueStore.cs:line 47
   at Octopus.Server.Extensibility.Authentication.Guest.Configuration.GuestConfigurationStore.MoveSettingsToDatabase()
   at Octopus.Server.Extensibility.Authentication.Guest.Configuration.GuestConfigurationStore.GetIsEnabled()
   at System.Linq.Enumerable.WhereArrayIterator`1.MoveNext()
   at Octopus.Server.OctopusServerEngine.InitializeExtensions() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Server\OctopusServerEngine.cs:line 87
   at Octopus.Server.OctopusServerEngine.Start() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Server\OctopusServerEngine.cs:line 69
   at Octopus.Server.Commands.RunCommand.Start() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Server\Commands\RunCommand.cs:line 48
   at Octopus.Shared.Startup.AbstractCommand.Octopus.Shared.Startup.ICommand.Start(String[] commandLineArguments, ICommandRuntime commandRuntime, OptionSet commonOptions, String displayName, String version, String informationalVersion, String[] environmentInformation, String instanceName) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\AbstractCommand.cs:line 78
   at Octopus.Shared.Startup.OctopusProgram.Start(ICommandRuntime commandRuntime) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\OctopusProgram.cs:line 252
   at Octopus.Shared.Startup.WindowsServiceHost.<>c__DisplayClass1_0.<Run>b__0() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\WindowsServiceHost.cs:line 20
   at Octopus.Shared.Startup.WindowsServiceAdapter.RunService() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\WindowsServiceAdapter.cs:line 44
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

### configure command
```
   at Octopus.Shared.Configuration.XmlFileKeyValueStore.AfterWriting() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Configuration\XmlFileKeyValueStore.cs:line 47
   at Octopus.Shared.Configuration.AbstractKeyValueStore.Set(String name, String value, Nullable`1 protectionScope) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Configuration\AbstractKeyValueStore.cs:line 80
   at Octopus.Core.WebPortal.WebPortalConfigurationStore.SetListenPrefixes(String listenPrefixes) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Core\WebPortal\WebPortalConfigurationStore.cs:line 39
   at Octopus.Server.Commands.ConfigureCommand.<>c__DisplayClass6_13.<.ctor>b__24() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Server\Commands\ConfigureCommand.cs:line 150
   at Octopus.Shared.Startup.AbstractMultiOperationCommand.Start() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\AbstractMultiOperationCommand.cs:line 17
   at Octopus.Server.Commands.ConfigureCommand.Start() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Server\Commands\ConfigureCommand.cs:line 188
   at Octopus.Shared.Startup.AbstractCommand.Octopus.Shared.Startup.ICommand.Start(String[] commandLineArguments, ICommandRuntime commandRuntime, OptionSet commonOptions, String displayName, String version, String informationalVersion, String[] environmentInformation, String instanceName) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\AbstractCommand.cs:line 78
   at Octopus.Shared.Startup.OctopusProgram.Start(ICommandRuntime commandRuntime) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\OctopusProgram.cs:line 252
   at Octopus.Shared.Startup.ConsoleHost.Run(Action`1 start, Action shutdown) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\ConsoleHost.cs:line 27
   at Octopus.Shared.Startup.OctopusProgram.Run() in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Shared\Startup\OctopusProgram.cs:line 101
   at Octopus.Server.Program.Main(String[] args) in Z:\buildAgent\workDir\eec88466c176b607\source\Octopus.Server\Program.cs:line 62
```

## Solution

We should prevent the 2 processes overwriting each other, potentially with a machine wide mutex.

> It probably should be supported (or at least not this dangerous) - we could employ a named `MachineWideMutex` so only one process can do configuration changes at a time. And we could do a bit of hash checking to make sure we’re working with a consistent view of the config before we save…

> Another approach would be to move away from the concept of groups of configuration values, and to individual values which update just themselves, like precision surgery… Or much smaller groups like `HSTSConfig` which has enough information to validate itself and save its own configuration.

**Comment by droyad** [comment]
@michaelnoonan Did you refactor fix this?

**Comment by michaelnoonan** [comment]
This world is slightly better now, but I think the race condition still exists under certain circumstances. In this instance I believe it's because the configure command is running concurrently with the Octopus Server starting up.

That being said, with the changes I've implemented in Octopus 3.15, I feel this issue could be closed: the scripts generated by the wizard create the database first, meaning the "lift-and-shift" of configuration settings is no longer a problem.

What I have done:
- Simplified the configuration command code
- Added a warning that the service requires restart after certain configuration changes
- Changed the script generated by the installation wizard so it sets up the database first removing the need to lift and shift configuration values from the configuration file to the database

What I haven't done:
- Added any kind of mutex or ordering enforcement to completely prevent this kind of issue recurring

**Comment by droyad** [comment]
☝️
